File: Behavioral/memento/memento.py
import copy

class Record:
    def __init__(self, data: str, nums: list):
        self.data = data
        self.nums = copy.deepcopy(nums)
        self.snapshots = list()

    def create_snapshot(self, id: int):
        self.snapshots.append(Memento(id, self.data, self.nums))

    def restore_last(self):
        if last := self.snapshots[-1]:
            self.data = last.data
            self.nums = copy.deepcopy(last.nums)

    def restore_by_id(self, id: int):
        for snapshot in self.snapshots:
            if snapshot.id == id:
                self.data = snapshot.data
                self.nums = copy.deepcopy(snapshot.nums)
                break


class Memento:
    def __init__(self, id: int, data: str, nums: list):
        self.id = id
        self.data = data
        self.nums = copy.deepcopy(nums)


record = Record("Hello World", [5, 20, 30])
record.create_snapshot(3)

record.data += "Another string"
record.nums.append(500)

record.create_snapshot(5)

record.data += "Last string"
record.nums.append(1000)
print(f"{record.data} {record.nums}")
record.restore_by_id(3)
print(f"{record.data} {record.nums}")
record.restore_last()
print(f"{record.data} {record.nums}")

File: Behavioral/memento/memento.rs
#[derive(Clone)]
pub struct Record {
    data: String,
    nums: Vec<u32>,
    snapshots: Vec<Memento>,
}

#[derive(Clone)]
struct Memento {
    id: u32,
    data: String,
    nums: Vec<u32>,
}

impl Record {
    pub fn new(data: String, nums: Vec<u32>) -> Self {
        Record{data, nums, snapshots: vec![]}
    }

    pub fn create_snapshot(&mut self, id: u32) {
        self.snapshots.push(Memento {
            id,
            data: self.data.clone(),
            nums: self.nums.clone(),
        })
    }

    pub fn restore_last(&mut self) {
        match self.snapshots.last() {
            Some(value) => {
                self.data = value.data.clone();
                self.nums = value.nums.clone();
            }
            None => {}
        }
    }

    pub fn restore_by_id(&mut self, id: u32) {
        match self.snapshots.iter().position(|x| x.id == id) {
            Some(index) => {
                self.data = self.snapshots[index].data.clone();
                self.nums = self.snapshots[index].nums.clone();
            }
            None => {}
        }
    }
}

fn main() {
    let mut data = Record::new(String::from("Start"), vec![10, 20 ,30]);
    data.create_snapshot(1);

    data.data.push_str("Additional data");
    data.nums.push(60);

    data.create_snapshot(3);

    data.data.push_str("Last data");
    data.nums.push(70);

    println!("{} {:?}", data.data, data.nums);

    data.restore_by_id(1);
    println!("{} {:?}", data.data, data.nums);

    data.restore_last();
    println!("{} {:?}", data.data, data.nums);
}
